### PR TITLE
Integration tests refactoring 

### PR DESCRIPTION
**Pull request by mohavro**
## Description

  - Refactored tests/integration/test_api_integration.py to use httpx.AsyncClient with env defaults, disabled rate
    limiting in tests, and dependency overrides for /token.
  - Fixed YAML/test issues and hardened workflow tests (local actions allowed, valid trigger set expanded, cache key
    exception), plus cleaned requirement validation.
  - Updated workflows: added pull_request trigger types in 42 files, fixed indentation issues, pinned opencode action
    ref, aligned pr-agent.yml to expected structure, and updated auto-assign documentation.
  - Codacy config clean-up by removing several language runtimes (Dart, Go, Java) and analysis tools (dartanalyzer, PMD, revive, lizard) that are not used in the repo.

  Tests

  - python -m pytest tests/integration/test_api_integration.py tests/integration/test_github_workflows.py tests/
    integration/test_github_workflows_helpers.py tests/integration/test_requirements_validation.py
  - Result: 3907 passed (warnings: passlib crypt deprecation, datetime.utcnow() deprecation, JWT key length).

  Notes

  - Many .github/workflows/*.yml files changed due to the added pull_request types block.
  - pr-agent.yml now includes the expected jobs and metadata to satisfy tests.

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [X] Code refactoring
- [ ] Performance improvement
- [ ] Test addition/update
- [ ] Dependency update

## Related Issues

Closes #

## Changes Made

-
-
-

## Testing

- [ ] Unit tests pass locally
- [ ] Integration tests pass locally
- [ ] Manual testing completed
- [ ] New tests added for new functionality

### Test Commands

```bash
# Commands used to test
pytest
npm test
```

### Code Quality

- [ ] My code follows the project's style guidelines
- [ ] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### Documentation

- [ ] I have updated the README.md (if needed)
- [ ] I have updated the CHANGELOG.md
- [ ] I have updated relevant documentation files
- [ ] I have added/updated docstrings

### Dependencies

- [ ] I have checked for dependency conflicts
- [ ] I have updated requirements.txt (if needed)
- [ ] I have updated package.json (if needed)

### Branch Management

- [ ] My branch is up to date with main
- [ ] I will delete this branch after merge
- [ ] This branch has a descriptive name
- [ ] This PR has been open for less than 2 weeks

## Additional Notes

**For Reviewers**: Please ensure the branch is deleted after merging this PR to maintain repository hygiene.

**Related Documentation**:
- [Contributing Guidelines](../CONTRIBUTING.md)
- [Branch Cleanup Guidelines](../BRANCH_CLEANUP_QUICK_REFERENCE.md)

---
## Summary by cubic
Refactors the FastAPI backend into modular routers with a dependency-injected, thread‑safe graph and a sanitized schema‑report surface, and hardens the schema_report CLI and CI to meet Linear Q‑DEV‑869. Adds richer tests, clearer docs, and cleans up visualizations and database wiring.

- **New Features**
  - /schema-report (md/html) with Bleach‑sanitized HTML and correct media types; export helpers for markdown/html/export_report.
  - Safer schema_report CLI with validated format enum, project‑root bootstrapping, stdout/file output, md/text/JSON, and robust logging; added run_tests.py to seed CI env.
  - Optional MCP tools (add_equity_node, graph://data/3d-layout); mcp/fastmcp added.

- **Refactors**
  - API: modular routers and models (new api/models), thread‑safe get_graph via DI; improved /api/health and generic 500s; auth/docstring cleanup; removed global SQLite memory and session_scope; stricter DB URL handling.
  - Reports/Visuals: tighter Markdown→HTML sanitization; new SchemaReportGenerator; simplified 2D/3D traces with validation and constant name updates; removed unused node trace.
  - Tests/CI/Deps: async httpx integration tests with rate limiting disabled; new data integration coverage; workflows add pull_request types and pin actions (opencode v0.1.0, CodeQL upload v4.30.7); Bandit excludes tests; YAML deps pinned (PyYAML==6.0, ruamel.yaml==0.18); CircleCI ensures JUnit dir.

<sup>Written for commit 283f25fd1080488d6ae622f39c69f369f3288e4c. Summary will update on new commits. <a href="https://cubic.dev/pr/DashFin-FarDb/financial-asset-relationship-db/pull/877">Review in cubic</a></sup>

### **Description**
This PR includes significant enhancements and refactoring across various test files:
- Refactored integration tests for GitHub Actions to improve YAML handling and assertions.
- Enhanced unit tests for API endpoints with better mock configurations and documentation.
- Added tests for development scripts to ensure proper configurations and ignore patterns.
- Introduced new helper functions to streamline test setups and improve readability.

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>test_github_workflows.py</strong><dd><code>Refactor GitHub Actions Workflow Tests</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

tests/integration/test_github_workflows.py
<li>Refactored YAML resolver handling for GitHub Actions.<br> <li> Improved workflow file retrieval logic.<br> <li> Enhanced assertions in workflow tests for better clarity.<br>

</details>

  </td>
  <td><a href="https://github.com/DashFin-FarDb/financial-asset-relationship-db/pull/877/files#diff-beb22b64961fc1bd1db2d23ffc9993259440be38824ffd024d4836d3ec9ac1a3">+242/-565</a></td>
</tr>
</table></td></tr><tr><td><strong>Tests
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>test_api.py</strong><dd><code>Improve Unit Tests for API Endpoints</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

tests/unit/test_api.py
<li>Added detailed docstrings to mock_graph fixture.<br> <li> Introduced apply_mock_graph helper for better mock configuration.<br> <li> Enhanced test structure with additional unit markers.<br>

</details>

  </td>
  <td><a href="https://github.com/DashFin-FarDb/financial-asset-relationship-db/pull/877/files#diff-3cabff7688a6836f0be39ae71c04a68b09858683669c2bf6d24281910999cb74">+343/-217</a></td>
</tr>

<tr>
  <td>
    <details>
      <summary><strong>test_dev_scripts.py</strong><dd><code>Enhance Development Scripts Testing</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

tests/unit/test_dev_scripts.py
<li>Added unit tests for Vercel and environment configurations.<br> <li> Verified .gitignore and documentation file structures.<br>

</details>

  </td>
  <td><a href="https://github.com/DashFin-FarDb/financial-asset-relationship-db/pull/877/files#diff-d289ea25567e3a8a1f8ed59aecc43ef98fc17f9d3e91b2e018286eba7fea085b">+19/-0</a>&nbsp; &nbsp; </td>
</tr>
</table></td></tr></tr></tbody></table>

> 💡 **Penify usage**:
>Comment `/help` on the PR to get a list of all available Penify tools and their descriptions

___

## **CodeAnt-AI Description**
Refactor and strengthen CI workflow and microagent tests

### What Changed
- Workflow validation tests rewritten for clearer, stricter checks: more precise error messages, additional allowed GitHub events, enforcement of job/step requirements (names, runs-on, steps, versions), secrets usage, caching, concurrency, timeout and permissions rules, and checks that action refs are pinned; pr-agent workflow specifically expected to use a "pr-agent-trigger" job and its steps are validated for tokens, checkout refs, Python/Node setup and versions.
- Action/version checks skip local/relative actions and now require an explicit @ref for external actions; checkout steps validate fetch-depth and recommend using runner-aware cache keys; deprecated GitHub syntax and dangerous run patterns are flagged.
- Microagent unit tests expanded and reorganized: added pytest markers, more frontmatter and body validations (semver, required fields, formatting, size and performance bounds), edge-case and regression tests, and additional documentation/format checks.
- Minor test imports and small API test fix to ensure test suite imports TestClient correctly.

### Impact
`✅ Clearer workflow failure messages`
`✅ Fewer flaky/false-positive CI test failures`
`✅ Earlier detection of malformed microagent files`
<details>
<summary><strong>💡 Usage Guide</strong></summary>

### Checking Your Pull Request
Every time you make a pull request, our system automatically looks through it. We check for security issues, mistakes in how you're setting up your infrastructure, and common code problems. We do this to make sure your changes are solid and won't cause any trouble later.

### Talking to CodeAnt AI
Got a question or need a hand with something in your pull request? You can easily get in touch with CodeAnt AI right here. Just type the following in a comment on your pull request, and replace "Your question here" with whatever you want to ask:
<pre>
<code>@codeant-ai ask: Your question here</code>
</pre>
This lets you have a chat with CodeAnt AI about your pull request, making it easier to understand and improve your code.

#### Example
<pre>
<code>@codeant-ai ask: Can you suggest a safer alternative to storing this secret?</code>
</pre>

### Preserve Org Learnings with CodeAnt
You can record team preferences so CodeAnt AI applies them in future reviews. Reply directly to the specific CodeAnt AI suggestion (in the same thread) and replace "Your feedback here" with your input:
<pre>
<code>@codeant-ai: Your feedback here</code>
</pre>
This helps CodeAnt AI learn and adapt to your team's coding style and standards.

#### Example
<pre>
<code>@codeant-ai: Do not flag unused imports.</code>
</pre>

### Retrigger review
Ask CodeAnt AI to review the PR again, by typing:
<pre>
<code>@codeant-ai: review</code>
</pre>

### Check Your Repository Health
To analyze the health of your code repository, visit our dashboard at [https://app.codeant.ai](https://app.codeant.ai). This tool helps you identify potential issues and areas for improvement in your codebase, ensuring your repository maintains high standards of code health.

</details>

### CODE REVIEW

File: tests/integration/test_requirements_validation.py [hunk 1]
```diff
@@ -18,60 +18,63 @@ class TestRequirementsDevChanges:
...
-        # Find all non-comment lines containing 'pyyaml'
-        pyyaml_lines = [l for l in lines if "pyyaml" in l.lower() and not l.strip().startswith("#")]
+        from packaging.requirements import Requirement
+
+        # Find all non-comment lines explicitly declaring PyYAML (ignore types-PyYAML)
+        pyyaml_lines = [
+            l
+            for l in lines
+            if l.strip() and not l.strip().startswith("#") and Requirement(l).name.lower() == "pyyaml"
+        ]
```
[the remark this answers is not in the document]
Comment (mohavro) [reply tests/integration/test_requirements_validation.py]: /q can you suggest code to fix this?

**Review by mohavro — COMMENTED** [review]
Resolve all outstanding comments and issues raised by reviewers and contributors. 

Where possible suggest code for fixing any issues identified. 

complete PR ready for merge

File: api/main.py [hunk 2]
```diff
@@ -585,483 +663,211 @@ async def get_all_relationships():
...
+        # Build edges directly from graph.relationships to avoid rebuilding
+        # from intermediate data structures. Only include edges where both
+        # source and target are in the asset_ids list.
+        asset_id_set = set(asset_ids)
+        for source_id in g.relationships:
+            if source_id not in asset_id_set:
+                continue
+            for target_id, rel_type, strength in g.relationships[source_id]:
+                if target_id in asset_id_set:
+                    edges.append(
```
[the remark this answers is not in the document]
Comment (mohavro) [reply api/main.py]:
```suggestion
        positions, asset_ids, asset_colors, _ = g.get_3d_visualization_data_enhanced()
        nodes: List[Dict[str, Any]] = []
        for i, asset_id in enumerate(asset_ids):
            asset = g.assets.get(asset_id)
            if asset is None:
                continue
            nodes.append(
                {
                    "id": asset_id,
                    "name": asset.name,
                    "symbol": asset.symbol,
                    "asset_class": asset.asset_class.value,
                    "x": float(positions[i, 0]),
                    "y": float(positions[i, 1]),
                    "z": float(positions[i, 2]),
                    "color": asset_colors[i],
                    "size": 5,
                }
            )

        edges: List[Dict[str, Any]] = []
        # Build edges directly from graph.relationships to avoid rebuilding
        # from intermediate data structures. Only include edges where both
        # source and target are in the asset_ids list.
        asset_id_set = set(g.assets.keys())
        for source_id in g.relationships:
            if source_id not in asset_id_set:
                continue
            for target_id, rel_type, strength in g.relationships[source_id]:
                if target_id in asset_id_set:
                    edges.append(
```

File: tests/unit/test_config_validation.py [hunk 3]
```diff
@@ -1,534 +1,964 @@
...
+        """
+        config_path = Path("frontend/tailwind.config.js")
+        assert config_path.exists(), "tailwind.config.js not found"
+
+        with open(config_path) as f:
+            return f.read()
+
+    @staticmethod
+    @staticmethod
+    def test_tailwind_config_exists():
```
[the remark this answers is not in the document]
Comment (mohavro) [reply tests/unit/test_config_validation.py]:
```suggestion
    @staticmethod
    def test_tailwind_config_exists():

```

File: tests/unit/test_snyk_workflow.py [hunk 4]
```diff
@@ -0,0 +1,492 @@
...
+        workflow_str = str(snyk_workflow).lower()
+        # Check for common secret patterns
+        forbidden_patterns = [
+            "password=",
+            "api_key=",
+            "access_token=",
+            "private_key",
+        ]
+        for pattern in forbidden_patterns:
+            assert pattern not in workflow_str
```
[the remark this answers is not in the document]
Comment (mohavro) [reply tests/unit/test_snyk_workflow.py]:
```suggestion
    def test_no_hardcoded_secrets(self, snyk_workflow_content: str) -> None:
        """Test that workflow contains no hardcoded secrets."""
        workflow_str = snyk_workflow_content.lower()
        # Check for common secret patterns
        forbidden_patterns = [
            "password: ",
            "api_key: ",
            "access_token: ",
        ]
        for pattern in forbidden_patterns:
            assert pattern not in workflow_str
```

File: src/reports/integration.py [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+    Generate a Markdown-formatted schema and metrics report for the provided asset relationship graph.
+
+    Parameters:
+        graph (AssetRelationshipGraph): The asset relationship graph to analyze.
+
+    Returns:
+        str: Markdown string containing the generated schema and metrics report.
+    """
+    generator = SchemaReportGenerator(graph)
+    return generator.generate()
```
[the remark this answers is not in the document]
Comment (mohavro) [reply src/reports/integration.py]:
```suggestion
from src.reports.schema_report import generate_schema_report


def generate_markdown_report(graph: AssetRelationshipGraph) -> str:
    """Thin wrapper around generate_schema_report for naming consistency."""
    return generate_schema_report(graph)
```

File: tests/integration/test_data_integration.py [hunk 6]
```diff
@@ -0,0 +1,576 @@
...
+        for asset in graph.assets.values():
+            repo.upsert_asset(asset)
+        session.commit()
+
+        # Verify assets were saved
+        saved_assets = repo.list_assets()
+        assert len(saved_assets) == len(graph.assets)
+
+        session.close()
+        engine.dispose()
```
[the remark this answers is not in the document]
Comment (mohavro) [reply tests/integration/test_data_integration.py]:
```suggestion
    @staticmethod
    def test_sample_graph_can_be_saved_to_repository(tmp_path):
        """Test that a sample graph can be saved to and loaded from repository."""
        from src.data.database import (
            create_engine_from_url,
            create_session_factory,
            init_db,
            )

        db_path = tmp_path / "test_integration.db"
        engine = create_engine_from_url(f"sqlite:///{db_path}")
        init_db(engine)
        factory = create_session_factory(engine)
        session = factory()

```

File: src/reports/schema_report_generator.py [hunk 7]
```diff
@@ -0,0 +1,294 @@
+from __future__ import annotations
+
+from typing import Any, Callable, Dict, Iterable, List, Tuple
```
[the remark this answers is not in the document]
Comment (mohavro) [reply src/reports/schema_report_generator.py]:
```suggestion
from typing import Any, Callable, Dict, Iterable, List
```

File: tests/unit/test_generate_status.py [hunk 8]
```diff
@@ -0,0 +1,930 @@
+#!/usr/bin/env python3
+"""
+Unit tests for .github/pr-copilot/scripts/generate_status.py
+
+Tests cover all functions, dataclasses, API interactions, and output formatting.
+"""
+
+from __future__ import annotations
+
+import os
```
[the remark this answers is not in the document]
Comment (mohavro) [reply tests/unit/test_generate_status.py]:
```suggestion
```

File: src/visualizations/graph_visuals.py [hunk 9]
```diff
@@ -1,163 +1,147 @@
...
+from typing import Dict, Optional
+
+import plotly.graph_objects as go
+
+from src.logic.asset_graph import AssetRelationshipGraph
+
+# Sub-module re-exports — kept here so existing imports from this module continue to work.
+from src.visualizations.graph_visuals_constants import REL_TYPE_COLORS  # noqa: F401
+from src.visualizations.graph_visuals_data import _build_asset_id_index  # noqa: F401
+from src.visualizations.graph_visuals_data import _build_edge_coordinates_optimized  # noqa: F401
```
[the remark this answers is not in the document]
Comment (mohavro) [reply src/visualizations/graph_visuals.py]:
```suggestion
```

File: src/api/routers/dependencies.py [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+_graph_lock = threading.Lock()
+
+
+def get_graph() -> AssetRelationshipGraph:
+    global _graph
+    if _graph is None:
+        with _graph_lock:
+            if _graph is None:
+                _graph = _initialize_graph()  # move _initialize_graph here too
+    return _graph
```
[the remark this answers is not in the document]
Comment (mohavro) [reply src/api/routers/dependencies.py]:
```suggestion
def get_graph() -> AssetRelationshipGraph:
    """
    Return the lazily initialised AssetRelationshipGraph singleton.

    Returns:
        AssetRelationshipGraph: The shared graph instance.

    Raises:
        RuntimeError: If initialisation fails (propagated from the graph builder).
    """
    global _graph
    if _graph is None:
        with _graph_lock:
            if _graph is None:
                _graph = _initialize_graph()  # move _initialize_graph here too
    return _graph
```

File: src/api/routers/schema_report.py [hunk 11]
```diff
@@ -0,0 +1,57 @@
+from __future__ import annotations
+
+from fastapi import APIRouter, Depends, HTTPException, Query
+from fastapi.responses import HTMLResponse, Response
+
+from src.api.dependencies import get_graph
+from src.logic.asset_graph import AssetRelationshipGraph
+from src.reports.integration import export_report, generate_html_report, generate_markdown_report
```
[the remark this answers is not in the document]
Comment (mohavro) [reply src/api/routers/schema_report.py]: can you suggest code to define the function in src/reports/integration.py?

File: src/data/repository.py [hunk 12]
```diff
@@ -21,6 +19,7 @@
     RegulatoryEvent,
 )
 
+from .database import session_scope  # noqa: F401
```
[the remark this answers is not in the document]
Comment (mohavro) [reply src/data/repository.py]:
```suggestion
```

File: api/main.py [hunk 13]
```diff
@@ -1,16 +1,25 @@
...
-for all workflow files in .github/workflows/
-"""
-
-import warnings as GLOBAL_WARNINGS
-from pathlib import Path
-from typing import Any, Dict, List
-
-import pytest
-import yaml
+# Comprehensive test coverage available in tests/unit/test_api_main.py
```
[the remark this answers is not in the document]
Comment (mohavro) [reply api/main.py]: Can you create a new issue detailing the requirements for this refactoring which should be addressed separately due to the task size?

File: tests/unit/conftest.py [hunk 14]
```diff
@@ -0,0 +1,14 @@
...
+
+os.environ.setdefault("DATABASE_URL", "sqlite:///:memory:")
+os.environ.setdefault("SECRET_KEY", "test-secret-key-for-ci")
+os.environ.setdefault("ADMIN_USERNAME", "admin")
+
+if "ADMIN_PASSWORD" not in os.environ:
+    pytest.skip(
+        "ADMIN_PASSWORD not set; skipping API tests that require auth.",
+        allow_module_level=True,
+    )
```
[the remark this answers is not in the document]
Comment (mohavro) [reply tests/unit/conftest.py]:
```suggestion
"""Unit test configuration — sets required environment variables before collection."""

import os
import sys  # noqa: F401 — retained if used by downstream imports; remove if not
import pytest


os.environ.setdefault("DATABASE_URL", "sqlite:///:memory:")
os.environ.setdefault("SECRET_KEY", "test-secret-key-for-ci")
os.environ.setdefault("ADMIN_USERNAME", "admin")

if "ADMIN_PASSWORD" not in os.environ:
    pytest.skip(
        "ADMIN_PASSWORD not set; skipping API tests that require auth.",
        allow_module_level=True,
    )
```

File: tests/unit/test_mcp_server.py [hunk 15]
```diff
@@ -0,0 +1,582 @@
...
+        # Calling add_asset with invalid data should raise ValueError, but the
+        # lock must still be released after the exception.
+        with pytest.raises(ValueError):
+            safe_graph.add_asset(None)
+
+        # Lock should not be held after exception
+        acquired = lock.acquire(blocking=False)
+        if acquired:
+            lock.release()
+        assert acquired, "Lock was not released after exception"
```
[the remark this answers is not in the document]
Comment (mohavro) [reply tests/unit/test_mcp_server.py]:
```suggestion
    def test_thread_safe_graph_with_exception_in_method():
        """Test that lock is released even if wrapped method raises exception."""
        from mcp_server import _ThreadSafeGraph

        graph = AssetRelationshipGraph()
        lock = threading.Lock()
        safe_graph = _ThreadSafeGraph(graph, lock)

        # Calling add_asset with invalid data should raise ValueError, but the
        # lock must still be released after the exception.
        with pytest.raises(AttributeError):
            safe_graph.add_asset(None)

        # Lock should not be held after exception
        acquired = lock.acquire(blocking=False)
        if acquired:
            lock.release()
        assert acquired, "Lock was not released after exception"

```

File: src/reports/integration.py [hunk 16]
```diff
@@ -0,0 +1,243 @@
...
+        protocols=_ALLOWED_PROTOCOLS,
+        strip=True,
+    )
+
+    # Add rel="nofollow noopener" to links and open in new tab defensively.
+    def _add_noopener(attrs, new=False):
+        rel = attrs.get((None, "rel"), "")
+        if "noopener" not in rel:
+            attrs[(None, "rel")] = (rel + " noopener").strip()
+        return attrs
```
[the remark this answers is not in the document]
Comment (mohavro) [reply src/reports/integration.py]:
```suggestion
    # Add rel="nofollow noopener" to links and open in new tab defensively.
    def _add_noopener(attrs: dict, new: bool = False) -> dict:
        rel = attrs.get((None, "rel"), "")
        if "noopener" not in rel:
            attrs[(None, "rel")] = (rel + " noopener").strip()
        return attrs
```

File: .github/scripts/schema_report_cli.py [hunk 17]
```diff
@@ -1,248 +1,371 @@
...
-    # Setup logging
-    logger = setup_logging(args.verbose)
-    logger.info("Schema Report CLI started")
+
+def format_as_json(metrics: Dict[str, Any]) -> str:
+    """
+    Format a JSON payload containing the provided Markdown schema report.
+
+    Args:
+        markdown: The Markdown schema report to include in the payload.
```
[the remark this answers is not in the document]
Comment (mohavro) [reply .github/scripts/schema_report_cli.py]:
```suggestion
def convert_markdown_to_json(metrics: Dict[str, Any]) -> str:
    """
    Format a JSON payload containing the provided Markdown schema report.

    Args:
        markdown: The Markdown schema report to include in the payload.
```

File: tests/unit/test_suggest_fixes.py [hunk 18]
```diff
@@ -0,0 +1,800 @@
...
+Unit tests for .github/pr-copilot/scripts/suggest_fixes.py
+
+Tests cover configuration loading, comment parsing, categorization,
+formatting, and report generation.
+"""
+
+from __future__ import annotations
+
+import sys
+import tempfile
```
[the remark this answers is not in the document]
Comment (mohavro) [reply tests/unit/test_suggest_fixes.py]:
```suggestion
```

File: .github/scripts/schema_report_cli.py [hunk 19]
```diff
@@ -1,248 +1,371 @@
...
+    fd, tmp_path_str = tempfile.mkstemp(dir=str(path.parent))
+    tmp_path = Path(tmp_path_str)
 
     try:
-        # Validate and convert format
+        with os.fdopen(fd, "w", encoding=encoding) as fh:
+            fh.write(data)
+            fh.flush()
+            os.fsync(fh.fileno())
+        os.replace(tmp_path, path)
```
Comment (mohavro) [inline .github/scripts/schema_report_cli.py]: ### <sub></sub>&nbsp; Path Traversal

Unsanitized input from `a command line argument` flows into `os.replace`, where it is used as a path. This may result in a Path Traversal vulnerability and allow an attacker to remove arbitrary files.

<h6>
Line 276 | <a href="https://cwe.mitre.org/data/definitions/23.html">CWE-23</a> | <a href="https://docs.snyk.io/manage-risk/prioritize-issues-for-fixing/priority-score">Priority score 608</a> | <a href="https://learn.snyk.io/lesson/directory-traversal/?utm_campaign=snyk-learn&utm_medium=inline-comment&utm_source=npx">Learn more about this vulnerability</h6>

<details>

<summary>Data flow: 20 steps</summary>

`Step 1 - 2` https://github.com/DashFin-FarDb/financial-asset-relationship-db/blob/aabf5deb25853c107c856369685e893844334cd2/.github/scripts/schema_report_cli.py#L210

`Step 3` [.github/scripts/schema_report_cli.py#L346](https://github.com/DashFin-FarDb/financial-asset-relationship-db/blob/aabf5deb25853c107c856369685e893844334cd2/.github/scripts/schema_report_cli.py#L346)

`Step 4 - 5` [.github/scripts/schema_report_cli.py#L352](https://github.com/DashFin-FarDb/financial-asset-relationship-db/blob/aabf5deb25853c107c856369685e893844334cd2/.github/scripts/schema_report_cli.py#L352)

`Step 6` [.github/scripts/schema_report_cli.py#L289](https://github.com/DashFin-FarDb/financial-asset-relationship-db/blob/aabf5deb25853c107c856369685e893844334cd2/.github/scripts/schema_report_cli.py#L289)

`Step 7` [.github/scripts/schema_report_cli.py#L316](https://github.com/DashFin-FarDb/financial-asset-relationship-db/blob/aabf5deb25853c107c856369685e893844334cd2/.github/scripts/schema_report_cli.py#L316)

`Step 8` [.github/scripts/schema_report_cli.py#L254](https://github.com/DashFin-FarDb/financial-asset-relationship-db/blob/aabf5deb25853c107c856369685e893844334cd2/.github/scripts/schema_report_cli.py#L254)

`Step 9 - 10` [.github/scripts/schema_report_cli.py#L266](https://github.com/DashFin-FarDb/financial-asset-relationship-db/blob/aabf5deb25853c107c856369685e893844334cd2/.github/scripts/schema_report_cli.py#L266)

`Step 11 - 15` [.github/scripts/schema_report_cli.py#L268](https://github.com/DashFin-FarDb/financial-asset-relationship-db/blob/aabf5deb25853c107c856369685e893844334cd2/.github/scripts/schema_report_cli.py#L268)

`Step 16 - 18` [.github/scripts/schema_report_cli.py#L269](https://github.com/DashFin-FarDb/financial-asset-relationship-db/blob/aabf5deb25853c107c856369685e893844334cd2/.github/scripts/schema_report_cli.py#L269)

`Step 19 - 20` https://github.com/DashFin-FarDb/financial-asset-relationship-db/blob/aabf5deb25853c107c856369685e893844334cd2/.github/scripts/schema_report_cli.py#L276

</details>

---

<details open>
<summary>Commands</summary>

- :zap: To see AI-powered Snyk Agent Fix suggestions, reply with: <b>`@snyk /fix`</b>. You'll need to refresh the page :arrows_counterclockwise:
</details>

File: .elastic-copilot/memory/systemManifest.md [hunk 20]
```diff
@@ -235,104 +243,128 @@ Dependencies:
...
 - react
 - @testing-library/react
 - @testing-library/jest-dom
 - ../../app/components/MetricsDashboard
 - ../../app/types/api
 - ../test-utils
 
-
-
 ## Project Directory Structure
```
[the remark this answers is not in the document]
Comment (mohavro) [reply .elastic-copilot/memory/systemManifest.md]: go ahead and create a new issue to track fixing the manifest generator's append-vs-replace behaviour.

File: tests/unit/test_schema_report_cli.py [hunk 21]
```diff
@@ -1,401 +1,140 @@
...
+    cli_src = Path(__file__).resolve().parents[2] / ".github" / "scripts" / "schema_report_cli.py"
+    assert cli_src.exists(), f"CLI source not found: {cli_src}"
+
+    cli_copy = tmp_path / "schema_report_cli_copy.py"
+    cli_copy.write_text(cli_src.read_text(encoding="utf-8"), encoding="utf-8")
+
+    # Import the copy so the module's top-level code runs with the patched env
+    monkeypatch.syspath_prepend(str(tmp_path))
+    module = importlib.import_module("schema_report_cli_copy")
+    importlib.reload(module)
```
[the remark this answers is not in the document]
Comment (mohavro) [reply tests/unit/test_schema_report_cli.py]:
```suggestion

    # Copy the real CLI script into tmp_path under a throwaway name
    cli_src = Path(__file__).resolve().parents[2] / ".github" / "scripts" / "schema_report_cli.py"
    assert cli_src.exists(), f"CLI source not found: {cli_src}"
    
    
    cli_copy = tmp_path / "schema_report_cli_copy.py"
    cli_copy.write_text(cli_src.read_text(encoding="utf-8"), encoding="utf-8")

    # Import the copy so the module's top-level code runs with the patched env
    monkeypatch.syspath_prepend(str(tmp_path))
    # Evict any cached module so import_module loads from this test's tmp_path.
    sys.modules.pop("schema_report_cli_copy", None)
    importlib.invalidate_caches()  # ensure the new tmp_path file is found
    module = importlib.import_module("schema_report_cli_copy")
    module = importlib.import_module("schema_report_cli_copy")
    return module
```

File: src/visualizations/graph_visuals.py [hunk 22]
```diff
@@ -1,163 +1,158 @@
...
+from typing import Dict, Optional
+
+import plotly.graph_objects as go
+
+from src.logic.asset_graph import AssetRelationshipGraph
+
+# Sub-module re-exports — kept here so existing imports from this module continue to work.
+from src.visualizations.graph_visuals_constants import REL_TYPE_COLORS  # noqa: F401
+from src.visualizations.graph_visuals_data import _build_relationship_index  # noqa: F401
+from src.visualizations.graph_visuals_data import _collect_and_group_relationships  # noqa: F401
```
[the remark this answers is not in the document]
Comment (mohavro) [reply src/visualizations/graph_visuals.py]:
```suggestion
```

**Comment by mohavro** [comment]
@codeant-ai Provide suggested code patches to relax specific assertions (give examples for ~5 spots)
